sdks/common/runner-api/src/main/proto/beam_runner_api.proto

```diff
@@ -0,0 +1,526 @@
+
+  // (Required) The DoFn for this ParDo.
+  FunctionSpec do_fn = 1;
+
+  // (Required) Additional pieces of context the DoFn may require that
+  // are not otherwise represented in the payload.
+  // (may force runners to execute the ParDo differently)
+  repeated DoFnContext required_context = 2;
+
+  // (Optional) A map from local input name to the specification for how to
+  // materialize and read it as a side input
+  map<string, SideInput> side_inputs = 3;
+
+  // (Optional) if the DoFn uses state, a map from id to the spec for the
+  // state cell.
+  map<string, StateSpec> state_specs = 4;
+
+  // (Optional) if the DoFn uses timers, a map from id to the spec for the
+  // timer.
+  map<string, TimerSpec> timer_specs = 5;
+
+  // Requirements on the context a DoFn might require that are not otherwise
+  // represented in the signature here.
+  enum DoFnContext {
+    WINDOW_ACCESS = 0;
+  }
+}
+
+
+message StateSpec {
+  // TODO: AST for state spec
+}
+
+message TimerSpec {
+  // TODO: AST for timer spec
+}
+
+enum IsBounded {
+  BOUNDED = 0;
+  UNBOUNDED = 1;
+}
+
+// The payload for the primitive Read transform.
+//
+// TODO: audit what needs to be in this payload
+message ReadPayload {
+
+  // (Required) The source for this Read
+  FunctionSpec source = 1;
+
+  // (Required) Whether the source is bounded or unbounded
+  IsBounded is_bounded = 2;
+
+  // TODO: consider other methods such as doesNotRequireSplitting() and producesSortedKeys()
+  // what else?
+}
+
+// The payload for the Window transform.
+//
+// TODO: consider making this non-primitive according to BEAM-<issue #>
+message WindowIntoPayload {
+  // (Required) the window fn for assigning windows
+  FunctionSpec window_fn = 1;
+}
+
+// The payload for the non-primitive Combine transform.
+message CombinePayload {
+
+  // (Required) The CombineFn for this Combine
+  FunctionSpec combine_fn = 1;
+
+  // (Required) A reference to the Coder to use for accumulators of the CombineFn
+  int64 accumulator_coder_ref = 2;
+
+  // (Optional) if the CombineFn uses side inputs, a map from the id to the spec
+  // for how to materialize and read the PCollection as a side input
+  map<string, SideInput> side_inputs = 3;
+}
+
+// A coder, the binary format for serialization and deserialization of data in
+// a pipeline.
+message Coder {
+
+  // (Required) A pipeline-scoped unique id for this node.
+  int64 id = 1;
+
+  // (Required) A cross-language, stable, unique identifier for the (possibly
+  // parametric) encoding.
+  string urn = 2;
+
+  // (Optional) If this coder is parametric, such as ListCoder(VarIntCoder),
+  // this is a list of the components. In order for encodings to be identical,
+  // the URN and all components must be identical, recursively.
+  repeated int64 component_coder_reference = 3;
+
+  // (Optional) An SDK-specific UDF implementing the coder.
+  //
+  // TODO: reconsider how to do this, aka CustomCoder
+  FunctionSpec custom_coder_fn = 4;
```

Comment:
It is a good question. They may simply be redundant because of the choice to re-use `FunctionSpec`. But perhaps the URN in the `Coder` message might be `urn:beam:coders:custom`, indicating that it is SDK-specific, and the URN in the `FunctionSpec` describes the SDK-specific coder.

I had intentionally thought to be more flexible - maybe something has _both_ a useful URN as well as an SDK-specific implementation. But let's say that my above paragraph is the intent.